### PR TITLE
[1.20.6] Add BlockTagIngredient for when a item tag cannot be used/doesn't exist

### PR DESCRIPTION
**Pull request by IThundxr**
In some certain use cases using the normal tag ingredient with a ItemTag isn't always possible as a tag for that can not exist. A good example of this is `minecraft:convertable_to_mud` this tag only exists as a block tag with no ItemTag that can be used instead.

Example usage: <https://github.com/Creators-of-Create/Create/blob/mc1.20.1/dev/src/generated/resources/data/create/recipes/mixing/mud_by_mixing.json>

The type id for this is "neoforge:block_tag"

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    protected final TagKey<Block> tag;
+
+    @Nullable
+    protected ItemStack[] itemStacks;
+
+    public BlockTagIngredient(TagKey<Block> tag) {
+        this.tag = tag;
+    }
+
+    protected void dissolve() {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:45]: This shouldn't be necessary, caching this is already handled by Ingredient. But is that cache accessible from the custom ingredient? Maybe not...

File: src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import javax.annotation.Nullable;
+import net.minecraft.core.Holder;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.tags.TagKey;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.block.Block;
+import net.neoforged.neoforge.common.NeoForgeMod;
+
+/**
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:25]: Javadoc needs a bit of formatting work (@link, @code, etc).

File: src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.crafting;
+
+import com.mojang.serialization.MapCodec;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java]: Prefer the JetBrains one.

File: src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java (hunk 1 above)
Comment (IThundxr) [reply src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:45]: I don't think i can access that cache from here

File: src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                }
+            }
+            itemStacks = list.toArray(ItemStack[]::new);
+        }
+    }
+
+    @Override
+    public Stream<ItemStack> getItems() {
+        dissolve();
+        return itemStacks != null ? Stream.of(itemStacks) : Stream.empty();
```
Comment (embeddedt) [inline src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java]: `itemStacks` will never be `null` after a call to `dissolve`

Comment (IThundxr) [reply src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java]: That's kinda of just to make intellij not complain, could replace it with a noinspection if you'd prefer that instead

**Comment by IThundxr** [comment]
todo: implement hashCode & equals

File: src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        dissolve();
+        if (itemStacks.length == 0) {
+            return stack.isEmpty();
+        }
+
+        for (ItemStack itemStack : itemStacks) {
+            if (itemStack != null && itemStack.is(stack.getItem())) {
+                return true;
+            }
+        }
```
Comment (KnightMiner) [inline src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java]: Why not just store a lazily initialized set of values to check against? Would be notably faster. Can use the set to construct `getItems()`

File: src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    @Override
+    public boolean test(@Nullable ItemStack stack) {
+        if (stack == null) {
+            return false;
+        }
+
+        dissolve();
+        if (itemStacks.length == 0) {
+            return stack.isEmpty();
+        }
```
Comment (KnightMiner) [inline src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java]: I know vanilla ingredients do this, but I don't think modded ingredients want this. Neo notably patches the vanilla ingredient to sync a dummy item to the client for empty ingredients to ensure an empty tag (e.g. unused mod compat) does not become "fill with nothing".

File: src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java (hunk 1 above)
Comment (KnightMiner) [reply src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:45]: Given this list is used in `test`, it might still be an issue as this cache is never invalidated. I suppose its only a problem if someone statically stores a block tag ingredient though.

File: .idea/icon.svg
Comment (Technici4n) [inline .idea/icon.svg:1]: Shouldn't this be left for another commit? :P

**Review by Technici4n — DISMISSED** [review]
Besides that `.idea/icon.svg` I'm happy. Maybe that can be sent in a separate PR alongside `.editorconfig`?

**Comment by IThundxr** [comment]
i don't really have much interest in maintaining a editorconfig pr, i'll make a separate one for the intellij icon

File: tests/src/main/java/net/neoforged/neoforge/debug/crafting/IngredientTests.java [hunk 7]
```diff
@@ -66,6 +68,37 @@
...
 public class IngredientTests {
+    @GameTest
+    @EmptyTemplate
+    @TestHolder(description = "Tests if BlockTagIngredient works")
+    static void blockTagIngredient(final DynamicTest test, final RegistrationHelper reg) {
+        reg.addProvider(event -> new RecipeProvider(event.getGenerator().getPackOutput(), event.getLookupProvider()) {
+            @Override
+            protected void buildRecipes(RecipeOutput output) {
+                ShapelessRecipeBuilder.shapeless(RecipeCategory.MISC, Items.MUD)
+                        .requires(new BlockTagIngredient(BlockTags.CONVERTABLE_TO_MUD).toVanilla())
```
Comment (Technici4n) [inline tests/src/main/java/net/neoforged/neoforge/debug/crafting/IngredientTests.java]: Sorry I missed this in my previous review: Can you wrap the ingredient in a `new TestEnabledIngredient(`? (See example below).

File: src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    @Override
+    public boolean test(@Nullable ItemStack stack) {
+        if (stack == null)
+            return false;
+
+        dissolve();
+        for (ItemStack itemStack : itemStacks) {
+            if (itemStack != null && itemStack.is(stack.getItem())) {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java]: This null check seems unnecessary?

**Review by Technici4n — CHANGES_REQUESTED** [review]

**Review by Technici4n — DISMISSED** [review]

File: src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            List<ItemStack> list = new ArrayList<>();
+            for (Holder<Block> block : BuiltInRegistries.BLOCK.getTagOrEmpty(tag)) {
+                ItemStack stack = new ItemStack(block.value());
+                if (!stack.isEmpty()) {
+                    list.add(stack);
+                }
+            }
+
+            if (list.isEmpty()) {
+                ItemStack itemStack = new ItemStack(net.minecraft.world.level.block.Blocks.BARRIER);
```
Comment (pupnewfster) [inline src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java]: This doesn't need to be a FQN because we are not in a patch.

**Review by pupnewfster — CHANGES_REQUESTED** [review]

**Review by pupnewfster — APPROVED** [review]
